### PR TITLE
Dhyanpathak/ch13/password authentication

### PR DESCRIPTION
**Pull request by dhyanpathak**
Uses BCrypt password encoder and has working sign in and logout functionality. 
This PR might have failed pipelines as I bet Travis CI is looking for JDBC_DATABASE_URL which is actually set in Heroku when postgres is added. 

If I don't wake up in time for this demo and y'all wanna push this to Heroku:
- add Postgres as a Heroku Add on
- merge this PR to master
- if it fails, delete the Postgres addon and rollback build

### CODE REVIEW

File: src/main/java/Sysc4806Group/demo/controllers/UserController.java [hunk 1]
```diff
@@ -1,34 +1,98 @@
...
+
+        Role userRole = roleRepository.findByName(Role.Roles.ROLE_USER)
+                .orElseThrow(() -> new RuntimeException("Error: Role is not found."));
+        roles.add(userRole);
+
+        if(hasAdminRole) {
+            Role adminRole = roleRepository.findByName(Role.Roles.ROLE_ADMIN)
+                    .orElseThrow(() -> new RuntimeException("Error: Role is not found."));
+            roles.add(adminRole);
+        }
```
Comment (SVD6) [inline src/main/java/Sysc4806Group/demo/controllers/UserController.java:87]: Nice

File: src/main/java/Sysc4806Group/demo/repositories/RoleRepository.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+import java.util.Optional;
+
+import Sysc4806Group.demo.entities.Role;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+
+@Repository
+public interface RoleRepository extends JpaRepository<Role, Long> {
```
Comment (SVD6) [inline src/main/java/Sysc4806Group/demo/repositories/RoleRepository.java:11]: Is a whole repo necessary for roles? Can't it just be a boolean?

**Review by SVD6 — CHANGES_REQUESTED** [review]
Add some comments in the controllers

File: src/main/java/Sysc4806Group/demo/repositories/RoleRepository.java (hunk 2 above)
Comment (dhyanpathak) [reply src/main/java/Sysc4806Group/demo/repositories/RoleRepository.java:11]: It's just 2 records for ROLE_USER and ROLE_ADMIN. What happens is that the tables are joined to create user_roles which reference PK of user and PK of role. Boolean isn't a scalable approach.

Comment (SVD6) [reply src/main/java/Sysc4806Group/demo/repositories/RoleRepository.java:11]: Sounds good

**Review by SVD6 — APPROVED** [review]
Looks good

**Pull request merged**